### PR TITLE
feat(ast): Creates CtTypeParameter which represents a type parameter.

### LINKED ISSUE #606: Distinguish between TypeParam declaration (types, methods) and TypeParamRefs
With ability to search by refs like getDeclaration().

Currently code analysis and manipulation capabilities are limited because it's impossible to detect type parameter shadowing.

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #606

### CODE REVIEW

File: src/test/java/spoon/support/visitor/java/JavaReflectionTreeBuilderTest.java [hunk 1]
```diff
@@ -76,12 +77,21 @@ public void testScannerAnnotation() throws Exception {
...
 		assertEquals("E extends java.lang.Comparable<? super E>", ctTypeParameterReference.toString());
 		assertEquals(1, ctTypeParameterReference.getBoundingType().getActualTypeArguments().size());
 		assertTrue(ctTypeParameterReference.getBoundingType().getActualTypeArguments().get(0) instanceof CtTypeParameterReference);
+
+		// New type parameter declaration.
+		assertEquals(1, aType.getFormalCtTypeParameters().size());
+		CtTypeParameter ctTypeParameter = aType.getFormalCtTypeParameters().get(0);
+		assertEquals("E extends java.lang.Comparable<? super E>", ctTypeParameter.toString());
+		assertEquals(1, ctTypeParameter.getSuperclass().getActualTypeArguments().size());
+		assertTrue(ctTypeParameter.getSuperclass().getActualTypeArguments().get(0) instanceof CtTypeParameterReference);
```
Comment (monperrus) [inline src/test/java/spoon/support/visitor/java/JavaReflectionTreeBuilderTest.java]: assert on ctTypeParameter.getSuperclass().getActualTypeArguments().get(0).toString()

File: src/test/java/spoon/test/annotation/AnnotationTest.java [hunk 2]
```diff
@@ -839,9 +854,9 @@ public void process(GlobalAnnotation annotation, CtElement element) {
 		}
 	}
 
-	class TypeAnnotationProcessor extends AbstractElementsProcessor<TypeAnnotation, CtElement> {
```
Comment (monperrus) [inline src/test/java/spoon/test/annotation/AnnotationTest.java]: undo this and remove scan getFormalTypeParameters

File: src/test/java/spoon/test/constructor/ConstructorTest.java [hunk 3]
```diff
@@ -104,13 +105,25 @@ public void testTypeAnnotationWithConstructorsOnFormalType() throws Exception {
...
+		assertEquals("T", typeParameter.getSimpleName());
+		assertEquals(1, typeParameter.getAnnotations().size());
+		assertIntersectionTypeInConstructor(typeParameter.getSuperclass());
+	}
 
-		assertNotNull(genericT.getBoundingType());
-		assertTrue(genericT.getBoundingType() instanceof CtIntersectionTypeReference);
-		CtIntersectionTypeReference<?> boundingType = genericT.getBoundingType().asCtIntersectionTypeReference();
+	private void assertIntersectionTypeInConstructor(CtTypeReference<?> boundingType1) {
+		assertNotNull(boundingType1);
```
Comment (monperrus) [inline src/test/java/spoon/test/constructor/ConstructorTest.java]: remove

File: src/test/java/spoon/test/parent/ParentContractTest.java [hunk 4]
```diff
@@ -131,13 +131,15 @@ public void testContract() throws Throwable {
...
+				setter.invoke(receiver, new Object[] { mockedArgument });
 				// we check that setParent has been called
 				verify(mockedArgument).setParent((CtElement) receiver);
 			} catch (AssertionError e) {
-				Assert.fail("call setParent contract failed for "+setter.toString()+" "+e.toString());
+				Assert.fail("call setParent contract failed for " + setter.toString() + " " + e.toString());
 			} catch (InvocationTargetException e) {
-				if (e.getCause() instanceof RuntimeException) {
+				if (e.getCause() instanceof UnsupportedOperationException) {
+					// ignore
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentContractTest.java]: boyscout rule: remove the hardcoded list above?

File: src/test/java/spoon/test/generics/GenericsTest.java [hunk 5]
```diff
@@ -515,6 +531,23 @@ public void testGenericsInConstructorCall() throws Exception {
...
+
+	public void testDeclarationOfTypeParameterReference() throws Exception {
+		CtType<Tacos> aTacos = buildClass(Tacos.class);
+		for (CtTypeParameterReference parameterReference : aTacos.getElements(new TypeFilter<CtTypeParameterReference>(CtTypeParameterReference.class) {
+			@Override
+			public boolean matches(CtTypeParameterReference element) {
+				return !(element instanceof CtWildcardReference) && !(element.getParent() instanceof CtReference);
+			}
+		})) {
+			assertNotNull(parameterReference.getDeclaration());
```
Comment (monperrus) [inline src/test/java/spoon/test/generics/GenericsTest.java]: stronger assert than NotNull

File: src/test/java/spoon/test/generics/GenericsTest.java [hunk 6]
```diff
@@ -515,6 +531,23 @@ public void testGenericsInConstructorCall() throws Exception {
...
-		assertEquals(3, buildClass(Paella.class).getElements(new TypeFilter<CtWildcardReference>(CtWildcardReference.class)).size());
+		assertEquals(3, buildClass(Paella.class).getElements(new TypeFilter<CtWildcardReference>(CtWildcardReference.class) {
+			@Override
+			public boolean matches(CtWildcardReference element) {
+				return element.getParent(CtTypeParameter.class) != null;
+			}
+		}).size());
+	}
+
+	public void testDeclarationOfTypeParameterReference() throws Exception {
```
Comment (monperrus) [inline src/test/java/spoon/test/generics/GenericsTest.java]: @test

**Comment by monperrus** [comment]
in all "old" add assert on typeParamRef.getDeclaration() and getTypeDeclaration()

File: src/main/java/spoon/reflect/declaration/CtFormalTypeDeclarer.java [hunk 7]
```diff
@@ -27,27 +27,57 @@
 public interface CtFormalTypeDeclarer extends CtElement {
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtFormalTypeDeclarer.java]: This abstract element defines a declaration that accepts formal type
 * parameters (aka generics), such as a CtType (<code>class A<E></code>), CtMethod or CtConstructor

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.declaration;
+
+import spoon.reflect.reference.CtTypeParameterReference;
+import spoon.reflect.reference.CtTypeReference;
+
+/**
+ * This element defines a declaration of a type parameter (aka generics).
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: For example, in class<E> { ... }, the "E" is modeled as an instance of CtTypeParameter.

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ */
+public interface CtTypeParameter extends CtType<Object> {
+	/**
+	 * Sets the superclass type.
+	 */
+	<C extends CtTypeParameter> C setSuperclass(CtTypeReference<?> superClass);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: move setSuperclass in CtType

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+	@Override
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: // override the return type

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	CtTypeParameterReference getReference();
+
+	@Override
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: // override the return type

File: src/test/java/spoon/support/visitor/java/JavaReflectionTreeBuilderTest.java (hunk 1 above)
Comment (GerardPaligot) [inline src/test/java/spoon/support/visitor/java/JavaReflectionTreeBuilderTest.java]: Done.

File: src/test/java/spoon/test/parent/ParentContractTest.java (hunk 4 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/parent/ParentContractTest.java]: Done.

File: src/test/java/spoon/test/generics/GenericsTest.java (hunk 5 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/generics/GenericsTest.java]: Asserts added on other tests.

File: src/main/java/spoon/reflect/declaration/CtFormalTypeDeclarer.java (hunk 7 above)
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/declaration/CtFormalTypeDeclarer.java]: Done.

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Pull request merged**